### PR TITLE
Charlesmchen/oversize text messages and arbitrary attachments

### PR DESCRIPTION
**Pull request by charlesmchen**
What's included:

* Modify Debug UI to send oversize text messages and attachments with an unknown mime type.
* Add generic attachment adapter to present arbitrary attachments in messages view.
* Present oversize text message as a snippet with indication to tap.

What's not included:

* Tap on oversize text message shows full message in separate view.
* Tap on arbitrary attachment shows ... share UI?

PTAL @michaelkirk 

### Incoming and outgoing oversize text messages and arbitrary attachments.

screen shot 2017-03-28 at 3 29 17 pm

screen shot 2017-03-28 at 3 29 12 pm

### CODE REVIEW

File: Signal/src/Models/TSMessageAdapaters/TSMessageAdapter.m [hunk 1]
```diff
@@ -131,7 +132,25 @@ - (instancetype)initWithInteraction:(TSInteraction *)interaction
...
+                        // TODO: Tune this value.
+                        const NSUInteger kMaxTextDisplayLength = 256;
+                        NSString *displayText = fullText;
+                        if (fullText.length > kMaxTextDisplayLength) {
+                            // Trim whitespace before _AND_ after slicing the snipper from the string.
+                            NSString *snippet =
+                                [[[fullText stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]]
+                                    substringWithRange:NSMakeRange(0, kMaxTextDisplayLength)]
+                                    stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
+                            displayText =
```
Comment (charlesmchen) [inline Signal/src/Models/TSMessageAdapaters/TSMessageAdapter.m]: Hmm, I should apply `DisplayableTextFilter`.

File: Signal/src/view controllers/DebugUITableViewController.h [hunk 2]
```diff
@@ -0,0 +1,14 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+#import <UIKit/UIKit.h>
+
+@class TSThread;
+
+@interface DebugUITableViewController : UITableViewController
```
Comment (michaelkirk) [inline Signal/src/view controllers/DebugUITableViewController.h]: see feedback from: https://github.com/WhisperSystems/Signal-iOS/pull/1898/files

File: Signal/src/view controllers/SignalAttachment.swift [hunk 3]
```diff
@@ -147,6 +156,12 @@ class SignalAttachment: NSObject {
     // Returns the file extension for this attachment or nil if no file extension
     // can be identified.
     var fileExtension: String? {
+        if dataUTI == SignalAttachment.kOversizeTextAttachmentUTI ||
+        dataUTI == SignalAttachment.kUnknownTestAttachmentUTI {
+            assert(false)
```
Comment (michaelkirk) [inline Signal/src/view controllers/SignalAttachment.swift]: nit: see also assertionFailure()

File: Signal/src/view controllers/SignalAttachment.swift [hunk 4]
```diff
@@ -147,6 +156,12 @@ class SignalAttachment: NSObject {
...
+            return ""
```
Comment (michaelkirk) [inline Signal/src/view controllers/SignalAttachment.swift]: nit: What's the distinction between returning "" and returning nil? It seems like a distinction which could surprise the caller.

File: Signal/src/Models/TSMessageAdapaters/TSGenericAttachmentAdapter.m [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+#import "TSGenericAttachmentAdapter.h"
+#import "AttachmentUploadView.h"
+#import "TSAttachmentStream.h"
+#import "UIColor+JSQMessages.h"
+#import "UIColor+OWS.h"
+#import "UIFont+OWS.h"
+#import <JSQMessagesViewController/JSQMessagesBubbleImage.h>
+#import <JSQMessagesViewController/JSQMessagesBubbleImageFactory.h>
+#import <MobileCoreServices/MobileCoreServices.h>
+#import <SignalServiceKit/MimeTypeUtil.h>
```
Comment (charlesmchen) [inline Signal/src/Models/TSMessageAdapaters/TSGenericAttachmentAdapter.m]: Done, here and in the other adapters.

File: Signal/src/Models/TSMessageAdapaters/TSGenericAttachmentAdapter.m [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+@interface TSGenericAttachmentAdapter ()
+
+@property (nonatomic) UIView *cachedMediaView;
```
Comment (charlesmchen) [inline Signal/src/Models/TSMessageAdapaters/TSGenericAttachmentAdapter.m]: Done. I also streamlined the other adapaters.

File: Signal/src/Models/TSMessageAdapaters/TSGenericAttachmentAdapter.h [hunk 7]
```diff
@@ -0,0 +1,14 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+#import "OWSMessageEditing.h"
+#import <JSQMessagesViewController/JSQMediaItem.h>
+
```
Comment (charlesmchen) [inline Signal/src/Models/TSMessageAdapaters/TSGenericAttachmentAdapter.h]: Dpme/

File: Signal/src/Models/TSMessageAdapaters/TSMessageAdapter.m [hunk 8]
```diff
@@ -131,7 +132,25 @@ - (instancetype)initWithInteraction:(TSInteraction *)interaction
 
                 if ([attachment isKindOfClass:[TSAttachmentStream class]]) {
                     TSAttachmentStream *stream = (TSAttachmentStream *)attachment;
-                    if ([stream isAnimated]) {
+                    if ([attachment.contentType isEqualToString:OWSMimeTypeOversizeTextMessage]) {
+                        NSData *textData = [NSData dataWithContentsOfURL:stream.mediaURL];
+                        NSString *fullText = [[NSString alloc] initWithData:textData encoding:NSUTF8StringEncoding];
+                        // TODO: Tune this value.
+                        const NSUInteger kMaxTextDisplayLength = 256;
```
Comment (charlesmchen) [inline Signal/src/Models/TSMessageAdapaters/TSMessageAdapter.m]: Done.

File: Signal/src/Models/TSMessageAdapaters/TSGenericAttachmentAdapter.h (hunk 7 above)
Comment (charlesmchen) [inline Signal/src/Models/TSMessageAdapaters/TSGenericAttachmentAdapter.h]: Done.

File: Signal/src/Models/TSMessageAdapaters/TSMessageAdapter.m [hunk 9]
```diff
@@ -143,13 +162,19 @@ - (instancetype)initWithInteraction:(TSInteraction *)interaction
...
                             initWithAttachment:stream
                                       incoming:[interaction isKindOfClass:[TSIncomingMessage class]]];
                         adapter.mediaItem.appliesMediaViewMaskAsOutgoing =
                             [interaction isKindOfClass:[TSOutgoingMessage class]];
                         break;
+                    } else {
+                        adapter.mediaItem = [[TSGenericAttachmentAdapter alloc]
+                            initWithAttachment:stream
+                                      incoming:[interaction isKindOfClass:[TSIncomingMessage class]]];
+                        adapter.mediaItem.appliesMediaViewMaskAsOutgoing = YES;
```
Comment (charlesmchen) [inline Signal/src/Models/TSMessageAdapaters/TSMessageAdapter.m]: Done and I simplified all of these.

File: Signal/src/util/ThreadUtil.h [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+
+@class TSThread;
+@class OWSMessageSender;
+@class SignalAttachment;
+
+@interface ThreadUtil : NSObject
```
Comment (charlesmchen) [inline Signal/src/util/ThreadUtil.h]: Done.

File: Signal/src/util/ThreadUtil.m [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+#import "ThreadUtil.h"
+#import "Signal-Swift.h"
+#import <SignalServiceKit/NSDate+millisecondTimeStamp.h>
+#import <SignalServiceKit/OWSDisappearingMessagesConfiguration.h>
+#import <SignalServiceKit/OWSMessageSender.h>
+#import <SignalServiceKit/TSOutgoingMessage.h>
+#import <SignalServiceKit/TSThread.h>
+
+@implementation ThreadUtil
```
Comment (charlesmchen) [inline Signal/src/util/ThreadUtil.m]: Done.

File: Signal/src/view controllers/DebugUITableViewController.h (hunk 2 above)
Comment (charlesmchen) [inline Signal/src/view controllers/DebugUITableViewController.h]: Done.

File: Signal/src/view controllers/SignalAttachment.swift (hunk 3 above)
Comment (charlesmchen) [inline Signal/src/view controllers/SignalAttachment.swift]: Done.

File: Signal/src/view controllers/SignalAttachment.swift (hunk 4 above)
Comment (charlesmchen) [inline Signal/src/view controllers/SignalAttachment.swift]: You're right, nil is better. done.

**Comment by charlesmchen** [comment]
@michaelkirk Thanks for the feedback.

**Pull request merged**